### DISCUSSION TITLE
Setting default file location

### DISCUSSION THREAD

**Discussion by putt1ck** [winder/Universal-G-Code-Sender#2634] [Help]
I have a workflow where the design work is done on one PC indoors and then a second "insecure" PC is used to connect to the CNC in the shed. To move files from the indoor machine to the outdoor one I have a read only share on a Nextcloud (NC) install; so on the indoor PC I save ugsd or gcode files to the appropriate local folder, which then syncs to NC, then on the outdoor one just download the files from the NC share.

This workflow would be neater at both ends if I could configure UGS to have a default file location, but it always opens in the user home directory - which is not a sensible place to keep files. I cannot see any where in the preferences, UI or CLI to set a default file location - is it set in the bundled Java?

**Comment by breiler** [comment]
Just to make sure nothing has been broken I tested this with the latest UGS version on Windows, Mac and Linux and it should work as expected.

Once you've opened a file the file dialog should remember the current directory and preselect it the next time. 

I know that there are some issues on Windows where you can create some kind of directory shortcut/link which will essentially be an odd proprietary file. If this is happening with Nextcloud as well could explain why it isn't working.

If you open the file "UniversalGcodeSender.json" from "c:/users/[your username]/.ugs" in a text editor it should have a list of the latest opened files. What is the path for the files stored on NC?

**Comment by putt1ck** [comment]
I cannot find that json in either ~/.ugs/ nor ~/.ugsplatform/.

All of my files are stored on NC (or more than one NC, depending on use case), replicated via profiles in NC desktop client. From desktop applications they only see the local files and they don't differ if they weren't being replicated to NC

**Reply by breiler** [reply]
> I cannot find that json in either ~/.ugs/ nor ~/.ugsplatform/.

Without that settings file it will be hard for me to help you. Maybe to log file will have some clues on the actual path for these files. You can find the log file by going to the menu "Tools -> Open log directory".

> All of my files are stored on NC (or more than one NC, depending on use case), replicated via profiles in NC desktop client. From desktop applications they only see the local files and they don't differ if they weren't being replicated to NC

This is built on the assumption that the NC desktop client is synchronizing actual files to the file system and is not providing its own virtual file system. The configuration file or log files would quickly answer this.

**Reply by putt1ck** [reply]
The comments about NC are built on the basis of my being involved in the project for since it forked from OwnCloud in 2016 :) (the local filesystem is unvarnished ext4, as is (in my install's case) the file system on the server; there's some other stuff it can do on Windows to deal with locks, but AFAIR it does it using Windows features).

Starting UGS the log (found in ~/.ugsplatform/var/log/) returns

`~  cat .ugsplatform/var/log/messages.log
-------------------------------------------------------------------------------
>Log Session: Thursday, November 7, 2024 at 2:12:43 PM Greenwich Mean Time
>System Info: 
  Product Version         = Universal Gcode Sender 18-7d33005615b3c744564f3d38ab9ef14bf8c3ec44
  Operating System        = Linux version 6.10.12-1 running on amd64
  Java; VM; Vendor        = 17.0.8.1; OpenJDK 64-Bit Server VM 17.0.8.1+1; Eclipse Adoptium
  Runtime                 = OpenJDK Runtime Environment 17.0.8.1+1
  Java Home               = /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/jdk/jdk-17.0.8.1+1-jre
  System Locale; Encoding = en_GB (ugsplatform); UTF-8
  Home Directory          = /home/chris
  Current Directory       = /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/bin
  User Directory          = /home/chris/.ugsplatform
  Cache Directory         = /home/chris/.cache/ugsplatform/var/cache
  Installation            = /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/extra
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/etc
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/ide
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/ugsplatform
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/platform
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/java
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/bin
                            /home/chris/ugsplatform-linux-x64/ugsplatform-linux-x64/platform
`

Which seems to suggest ~/.ugsplatform is where the UniversalGcodeSender.json file should be located. The log references the file though, just once just after startup: `INFO [com.willwinder.universalgcodesender.utils.SettingsFactory]: Log location: /home/chris/.ugs/UniversalGcodeSender.json`

**Reply by breiler** [reply]
> The comments about NC are built on the basis of my being involved in the project for since it forked from OwnCloud in 2016

Allrighty then! =)

> Which seems to suggest ~/.ugsplatform is where the UniversalGcodeSender.json file should be located.

The `~/.ugsplatform` directory contains a bunch of other settings for the UGS platform and not the `UniversalGcodeSender.json`. The ~/.ugs/ directory should (as the log file hints about) contain this setting file. If it doesn't exist it is likely the cause of your problem. This file contains a small history of the latest opened files which is used by the "File Open"-dialog.

In that case I would like to have your log file after you've closed UGS.

**Comment by putt1ck** [comment]
I did `touch .ugs/UniversalGcodeSender.json` and the file now gets content on exit. So looking at the content I see `"lastWorkingDirectory": "/home/chris" so presumably I could modify that to be the directory I would normally work in.

**Reply by breiler** [reply]
Since you insist on not sending me either the logs or setting file like I've asked for three times I will have to continue wasting time on guessing.

The fact that you needed to create the file is in it self a problem which I could have got answers to if you sent me the log. But now the chance for me to figure that out is probably gone.

It does not use the 'lastWorkingDirectory', instead it will use the first entry in `fileHistory` which contains a list of the latest files that you have opened. The settings will only reside in memory until you close UGS which will then be written to the settings file.
